### PR TITLE
Implement sprinting mechanics (2nd try)

### PR DESCRIPTION
**Pull request by CI09**
**Brief Description of What This PR Does**

Ports @ender-titan1 's sprint PR to newer version. Currently many features are missing or are malformed but can be still brought back to a good state

**Related Issues**

closes #4801

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by CI09** [comment]
It sort of can be initially checked upon, but know that it is quite rough port.

File: src/microbe_stage/PlayerMicrobeInput.cs [hunk 1]
```diff
@@ -303,6 +303,41 @@ public void CloseSignalingCommandsMenu()
             stage.HUD.ApplySignalCommand(command, stage.Player);
     }
 
+    [RunOnKeyDown("g_sprint")]
+    public bool StartSprint()
+    {
+        // Maybe add check for experimental features
```
Comment (hhyyrylainen) [inline src/microbe_stage/PlayerMicrobeInput.cs:309]: Yes, should add the experimental features check here.

File: src/microbe_stage/PlayerMicrobeInput.cs [hunk 2]
```diff
@@ -303,6 +303,41 @@ public void CloseSignalingCommandsMenu()
...
+            return;
+
+        ref var control = ref stage.Player.Get<MicrobeControl>();
+
+        control.Sprinting = false;
+    }
+
+    public void ToggleSprint()
+    {
+        if (!stage.HasPlayer)
```
Comment (hhyyrylainen) [inline src/microbe_stage/PlayerMicrobeInput.cs:333]: And here as well.

File: src/microbe_stage/components/MicrobeControl.cs [hunk 3]
```diff
@@ -64,6 +64,11 @@ public struct MicrobeControl
     /// </summary>
     public bool SlowedBySlime;
 
+    /// <summary>
+    ///   Whether this microbe is currenyly sprinting
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/MicrobeControl.cs:68]:
```suggestion
    ///   Whether this microbe is currently sprinting
```

File: src/microbe_stage/components/StrainAffected.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+[JSONDynamicTypeAllowed]
+public struct StrainAffected
+{
+    /// <summary>
+    ///   The current amount of strain
+    /// </summary>
+    public float CurrentStrain;
+
+    /// <summary>
+    ///   The amount of time the player has to wait before <see cref="CurrentStrain"/> sarts to fall
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/StrainAffected.cs:15]: This shouldn't explicitly talk about the player as other cells when the AI is updated should be able to sprint

File: src/microbe_stage/systems/StrainSystem.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+using DefaultEcs;
+using DefaultEcs.System;
+using DefaultEcs.Threading;
+using Godot;
+using World = DefaultEcs.World;
+
+[With(typeof(OrganelleContainer))]
+[With(typeof(StrainAffected))]
+[ReadsComponent(typeof(OrganelleContainer))]
+[ReadsComponent(typeof(MicrobeControl))]
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:11]: This seems like this system doesn't actually read this component...

File: src/microbe_stage/systems/StrainSystem.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+public sealed class StrainSystem : AEntitySetSystem<float>
+{
+    public StrainSystem(World world, IParallelRunner runner) : base(world, runner)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:14]:
```suggestion
    public StrainSystem(World world, IParallelRunner runner) : base(world, runner, Constants.SYSTEM_EXTREME_ENTITIES_PER_THREAD)
```

I'd add that to prepare this system to be multithreaded in the future.

File: src/microbe_stage/systems/StrainSystem.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    {
+    }
+
+    protected override void Update(float delta, in Entity entity)
+    {
+        ref var strain = ref entity.Get<StrainAffected>();
+        ref var organelles = ref entity.Get<OrganelleContainer>();
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:21]: This should only be fetched in the if condition where used to increase performance of this system.

File: src/microbe_stage/systems/StrainSystem.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            strain.CurrentStrain += strainIncrease;
+
+            if (strain.CurrentStrain > Constants.MAX_STRAIN_PER_ENTITY)
+                strain.CurrentStrain = Constants.MAX_STRAIN_PER_ENTITY;
+
+            strain.StrainDecreaseCooldown = Constants.STRAIN_DECREASE_COOLDOWN_SECONDS;
+        }
+        else
+        {
+            if (strain.StrainDecreaseCooldown <= Mathf.Epsilon)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:37]: Can't this variable go negative as well? It should be fine to just compare against zero in that case as I think it'll be more performant than comparing against an epsilon value unnecessarily.

File: src/microbe_stage/systems/StrainSystem.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            }
+            else
+            {
+                strain.StrainDecreaseCooldown -= delta;
+                ReduceStrain(ref strain, delta, Constants.PASSIVE_STRAIN_DECREASE_PRE_COOLDOWN_DIVISOR);
+            }
+        }
+    }
+
+    private void ReduceStrain(ref StrainAffected strain, float delta, float divisor = 1.0f)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:49]: Should use a multiplier instead of divisor.

File: project.godot [hunk 10]
```diff
@@ -491,6 +491,11 @@ g_eject_engulfed={
 "events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":72,"physical_keycode":0,"key_label":0,"unicode":0,"echo":false,"script":null)
 ]
 }
+g_sprint={
```
Comment (hhyyrylainen) [inline project.godot:494]: This new key must be added to the inputs json file so that it is rebindable.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I think fundamentally this feature is fine but I see quite a lot of commented out code and some TODO comments about unfinished parts in this PR. This is basically a full review of everything I think is still missing or non-optimal, overall the architecture doesn't need to change so all the changes I'm asking for are pretty minor.

File: src/general/base_stage/CreatureStageHUDBase.cs [hunk 11]
```diff
@@ -712,6 +719,63 @@ protected virtual void ReadPlayerHitpoints(out float hp, out float maxHP)
...
+
+        switch (Settings.Instance.StrainBarVisibilityMode.Value)
+        {
+            // TODO: second and thrid setting handling
+
+            case Settings.StrainBarVisibility.Off:
+                strainBar.Hide();
+                break;
+            case Settings.StrainBarVisibility.VisibleWhenCloseToFull:
+                if (strainFraction >= 0.8f)
```
[the remark this answers is not in the document]
Comment (CI09) [reply src/general/base_stage/CreatureStageHUDBase.cs:750]: Like this?
image

[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply src/general/base_stage/CreatureStageHUDBase.cs:750]: Well that's the preferred style for non-simple case statements, but the styleguide should say that this is not allowed:
```
if ()
   stuff
else
   other stuff
```
Instead multiline ifs must always have braces:
```
if ()
{
   stuff
}
else
{
   other stuff
}
```

File: src/microbe_stage/components/StrainAffected.cs (hunk 4 above)
Comment (CI09) [reply src/microbe_stage/components/StrainAffected.cs:15]: Should AI be updated in seperate PR to keep things clear or pack it here for consistency?

File: src/microbe_stage/systems/StrainSystem.cs (hunk 6 above)
Comment (CI09) [reply src/microbe_stage/systems/StrainSystem.cs:14]: It tells me I need to declare content when I add this, I'm kinda confused

File: src/microbe_stage/components/StrainAffected.cs (hunk 4 above)
Comment (hhyyrylainen) [reply src/microbe_stage/components/StrainAffected.cs]: I've let a bunch of features through that are missing the AI component, so I will let this PR go as well even without AI updates.

File: src/microbe_stage/MicrobeHUD.tscn [hunk 12]
```diff
@@ -712,6 +737,7 @@ layout_mode = 1
 [connection signal="pressed" from="ScrollContainer/HotBar/BindingMode" to="." method="OnBindingModePressed"]
 [connection signal="pressed" from="ScrollContainer/HotBar/UnbindAll" to="." method="OnUnbindAllPressed"]
 [connection signal="pressed" from="ScrollContainer/HotBar/SecreteSlime" to="." method="OnSecreteSlimePressed"]
+[connection signal="pressed" from="ScrollContainer/HotBar/Sprint" to="." method="OnFireToxinPressed"]
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.tscn:740]: The bug here is probably: `OnFireToxinPressed`

**Review by hhyyrylainen — APPROVED** [review]
This should finally be good enough to be merged as an experimental feature.

**Comment by hhyyrylainen** [comment]
Our git styleguide calls for PRs with multiple authors to do a manual squash to preserve author information, however this branch has intermittent merge commits so it causes a ton of problems when trying to manually squash this into a few commits, so I will do the "wrong" thing and just squash merge this whole thing as one commit.

**Comment by hhyyrylainen** [comment]
I created some follow up issues: https://github.com/Revolutionary-Games/Thrive/issues/5284

**Pull request merged**